### PR TITLE
WIP: Centralize in DeviceObserver log messages on device add/del

### PR DESCRIPTION
**Pull request by Cupidazul**
Centralize in DeviceObserver log messages when device is added or deleted:

Notes:
 * switched from db* functions to eloquent.
 * added  public functions:  
         * created
         * deleted
* added log::event in deleting also.

I don't know why, but the log::event inside the "deleting" doesn't work sometimes, that's why I added the "deleted" function, and then It worked.

Messages format:
created:
``` "Device $device->hostname has been created"  ```
deleted:
``` "Device $device->hostname has been deleted"  ```
deleting:
``` "Device $device->hostname has been removed"  ```

This way, if the community reports any problem, we know which message came through and from what function.
We just don't want to loose the message on eventlog, just because the device doesn't exists inside the db...
As this Observer is used in other parts of LibreNMS, I wanted to be sure we allways got eventlog with the information that "some librenms user deleted a certain device", this is important to track changes, this is not a full Audit, but a basic nice to have feature for the community...

Thank you @murrant and @Jellyfrog for all the help, and teamwork!
Any needfull, just let me know...
cheers!

Here are the usual tests/checks:
```
~$ ./lnms dev:check
Running PHP lint check...
PHP 7.4.23 | 10 parallel jobs
..                                                           2/2 (100 %)


Checked 2 files in 0.1 seconds
No syntax error found
success (0.13s)
Running style check...
PHP CS Fixer 2.19.0 Testament by Fabien Potencier and Dariusz Ruminski
Runtime: PHP 7.4.23
Loaded config default from ".php-cs-fixer.php".
Using cache file ".php_cs.cache".
Paths from configuration file have been overridden by paths provided as command arguments.
FS
Legend: ?-unknown, I-invalid file syntax (file ignored), S-skipped (cached or empty file), .-no changes, F-fixed, E-error
   1) app/Observers/DeviceObserver.php (no_whitespace_in_blank_line)

Fixed all files in 0.141 seconds, 18.000 MB memory used
success (0.63s)
Running unit check...
PHPUnit 9.5.6 by Sebastian Bergmann and contributors.

SSSSSSSSSSSS....SSSSSSSSSSSS........S.........S......SSSSSSSS   61 / 2479 (  2%)
.SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS  122 / 2479 (  4%)
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS  183 / 2479 (  7%)
SSSSSSSSSSSSSSSSSSSS.........................................  244 / 2479 (  9%)
.............................................................  305 / 2479 ( 12%)
.............................................................  366 / 2479 ( 14%)
.............................................................  427 / 2479 ( 17%)
.............................................................  488 / 2479 ( 19%)
.............................................................  549 / 2479 ( 22%)
.............................................................  610 / 2479 ( 24%)
.............................................................  671 / 2479 ( 27%)
.............................................................  732 / 2479 ( 29%)
.............................................................  793 / 2479 ( 31%)
.......................................................SS....  854 / 2479 ( 34%)
....................................SSSSSSS..........SSSSSSSS  915 / 2479 ( 36%)
SS.............SSSSSSSS......................................  976 / 2479 ( 39%)
.........

Time: 04:44.163, Memory: 446.50 MB

OK, but incomplete, skipped, or risky tests!
Tests: 2479, Assertions: 4602, Skipped: 202.
success (289.38s)
```

As we saw, me and @Jellyfrog , there are some occurrences of lost eventlogs that might occur in this situations:

 * polling has started for the device.  and at the same time "Device delete! " is requested, polling will add eventlogs of
what interfaces and updates he might encounter, this might occur after device was deleted (where eventlog was cleared).

For this we should add, to the daily cleanup scripts, a sanitization SQL:
```delete from eventlog where device_id not in (select device_id from devices);```

This delete, will not delete log::event generated inside function "DeviceObserver.deleting", because here we forced device=null .

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: app/Observers/DeviceObserver.php [hunk 1]
```diff
@@ -6,6 +6,28 @@
...
 {
+    /**
+     * Handle the device "created" event.
+     *
+     * @param  \App\Models\Device  $device
+     * @return void
+     */
+    public function created(Device $device)
+    {
+        \Log::event("Device $device->hostname has been created", $device, 'system', 3);
```
Comment (Jellyfrog) [inline app/Observers/DeviceObserver.php:17]: add a `use` for this and remove the `\`

File: app/Models/Port.php [hunk 2]
```diff
@@ -61,30 +61,34 @@ public static function boot()
...
     {
-        $os = $this->device->os;
+        $label = '';
 
-        if (\LibreNMS\Config::getOsSetting($os, 'ifname')) {
-            $label = $this->ifName;
-        } elseif (\LibreNMS\Config::getOsSetting($os, 'ifalias')) {
-            $label = $this->ifAlias;
-        }
+        if (isset($this->device) && is_object($this->device)) {
```
Comment (Jellyfrog) [inline app/Models/Port.php:66]: This is wrong, it prevents the code that's not dependent on the device to run.
The code I posted on discord does the correct thing.

File: includes/functions.php [hunk 3]
```diff
@@ -340,21 +340,24 @@ function delete_device($id)
     foreach ($fields as $field) {
         foreach (dbFetch('SELECT TABLE_NAME FROM information_schema.columns WHERE table_schema = ? AND column_name = ?', [$db_name, $field]) as $table) {
             $table = $table['TABLE_NAME'];
+            if ($table == 'devices') {
+                continue;
+            }
```
Comment (Jellyfrog) [inline includes/functions.php:345]: Why this?

Comment (Jellyfrog) [reply includes/functions.php:345]: Nvm.

File: LibreNMS/Util/Url.php [hunk 4]
```diff
@@ -139,7 +139,7 @@ public static function portLink($port, $text = null, $type = null, $overlib = tr
             $text = $label;
         }
 
-        $content = '<div class=list-large>' . addslashes(htmlentities($port->device->displayName() . ' - ' . $label)) . '</div>';
+        $content = '<div class=list-large>' . addslashes(htmlentities((isset($port->device) ? $port->device->displayName() : '') . ' - ' . $label)) . '</div>';
```
Comment (Jellyfrog) [inline LibreNMS/Util/Url.php:142]: you should split this to its own PR

File: app/Models/Port.php (hunk 2 above)
Comment (Jellyfrog) [reply app/Models/Port.php:66]: This should be in a own pr together with the changes above

File: LibreNMS/Util/Url.php (hunk 4 above)
Comment (murrant) [reply LibreNMS/Util/Url.php:142]: Additionally, calling portLink on the port of a deleted device is not going to be useful

File: app/Models/Port.php (hunk 2 above)
Comment (murrant) [reply app/Models/Port.php:66]: Again, we shouldn't have ports that don't have a device.

File: app/Observers/DeviceObserver.php [hunk 5]
```diff
@@ -38,6 +61,9 @@ public function updated(Device $device)
      */
     public function deleting(Device $device)
     {
+        // delete with NULL Device, so it doesn't get's deleted by daily cleanup scripts
+        Log::event("Device $device->hostname has been removed", null, 'system', 3);
```
Comment (murrant) [inline app/Observers/DeviceObserver.php:65]: deleting and deleted are both called when deleting a model.  So you will get two eventlog entries.

**Comment by murrant** [comment]
Remove all of the port changes and figure out why you have left over ports in your database after deleting a device.

**Comment by Cupidazul** [comment]
> Remove all of the port changes and figure out why you have left over ports in your database after deleting a device.

As we saw, me and @Jellyfrog , there are some occurrences of lost eventlogs that might occur in this situations:

polling has started for the device. and at the same time "Device delete! " is requested, polling will add eventlogs of
what interfaces and updates he might encounter, this might occur after device was deleted (where eventlog was cleared).

For this we should add, to the daily cleanup scripts, a sanitization SQL:
delete from eventlog where device_id not in (select device_id from devices);

This delete, will not delete log::event generated inside function "DeviceObserver.deleting", because here we forced device=null .

File: app/Models/Port.php (hunk 2 above)
Comment (Cupidazul) [reply app/Models/Port.php:66]: > This is wrong, it prevents the code that's not dependent on the device to run.
> The code I posted on discord does the correct thing. 
> 

Sorry for this, I must saw your code. I'll right on to it...

Comment (Cupidazul) [reply app/Models/Port.php:66]: > This should be in a own pr together with the changes above

Ok, let's solidify this code and all fixes, and I'll separate this into two new PRs.

File: LibreNMS/Util/Url.php (hunk 4 above)
Comment (Cupidazul) [reply LibreNMS/Util/Url.php:142]: Ok, let's solidify this code and all fixes, and I'll separate this into two new PRs.

File: app/Observers/DeviceObserver.php (hunk 5 above)
Comment (Cupidazul) [reply app/Observers/DeviceObserver.php:65]: ok, will remove log event from deleting, but, If you wish we can do it the other way around.
thx for the help.

**Comment by murrant** [comment]
Ok, so step one before a delete is set the device to disabled... Then start deleting.

**Comment by Cupidazul** [comment]
> Ok, so step one before a delete is set the device to disabled... Then start deleting.

:) Sorry I didn't understand... If we delete the device the disable will disapear, right?
You want me to change anything ?

**Comment by Cupidazul** [comment]
closing this PR. thx.
